File: ts/day2/day2.ts
/**
*  @File    :   day2.ts
*  @Time    :   2025/12/02 14:24:00
*  @Author  :   Alejandro Marrero (amarrerd@ull.edu.es)
*  @Version :   1.0
*  @Contact :   amarrerd@ull.edu.es
*  @License :   (C)Copyright 2025, Alejandro Marrero
*  @Desc    :   None
**/


import * as fs from 'fs';
import * as path from 'path';
import { start } from 'repl';

function readInput(fname: string = 'test.in'): [number, number][] {
    let ranges :  [number, number][] = [];
    const file_path = path.join(__dirname, fname);
    const lines = fs.readFileSync(file_path, 'utf-8').split(',');

    for (const line of lines) {
        const [start, end] = line.split('-');
        ranges.push([+start, +end]);

    }
    return ranges;
}


function invalids(start: number, end: number, regex: RegExp = /(.+)\1/): number {
    let total = 0;
    for (let i = start; i <= end; i++){
        const s = i.toString();
        const match = s.match(regex);
        if (match && match[0].length == s.length) {
            total += i;
        }
    }
    return total;
}




const ranges = readInput('input.in');
const part_one = ranges.reduce((acc, [start, end]) => acc + invalids(start, end), 0);
const re_part_two: RegExp = /^(.+)\1+$/
const part_two = ranges.reduce((acc, [start, end]) => acc + invalids(start, end, re_part_two), 0);
console.log(part_one, part_two);

File: py/day2/day2.py
#!/usr/bin/env python
# -*-coding:utf-8 -*-
"""
@File    :   day2.py
@Time    :   2025/12/02 13:24:36
@Author  :   Alejandro Marrero (amarrerd@ull.edu.es)
@Version :   1.0
@Contact :   amarrerd@ull.edu.es
@License :   (C)Copyright 2025, Alejandro Marrero
@Desc    :   None
"""

from pathlib import Path
import re
from typing import List, Tuple


def parse_input(fname: str = "test.in") -> List[Tuple[int, int]]:
    ranges = []
    with open(Path(__file__).with_name(fname), "r") as file:
        content = file.readline().split(",")
    for rid in content:
        start, end = rid.split("-")
        ranges.append((int(start), int(end)))

    return ranges


def invalids(start: int, end: int, regex: str = r"(.+)\1") -> int:
    total = 0
    for i in range(start, end + 1):
        s = str(i)
        match = re.match(regex, s)
        if match and len(match.group(0)) == len(s):
            total += i

    return total


if __name__ == "__main__":
    ranges = parse_input(fname="input.in")
    part_one: int = sum(invalids(start, end) for start, end in ranges)
    re_two = r"^(.+)\1+$"
    part_two: int = sum(invalids(start, end, regex=re_two) for start, end in ranges)
    print(part_one, part_two)
